app/views/advanced_messenger/_body_bottom.html.erb

```diff
@@ -0,0 +1,170 @@
+                            }]
+    %>
+
+    <%issues_or_forum_data.each do |issues_or_forum| %> 
+        <% if issues_or_forum[:grouped_unread_notifications].count == 0%>
+            <%= javascript_tag do %>
+                $("#notifications-placeholder").append(
+                    "<h3 style='color:#2A5685'><%=t(issues_or_forum[:my_page_all_notifications_read_text])%></h3>"
+                );
+            <%end%>
+        <%else%>
+            <%= javascript_tag do %>
+                html = "<h3 style='color:#2A5685'><%=t(issues_or_forum[:my_page_unread_notifications_text])%></h3><ul style='list-style-type: none; padding-inline-start: 15px; font-weight: bold;'>";
+                <%issues_or_forum[:grouped_unread_notifications].each_value do |unread_notification_status| %>  
+                    html += "<li>" + 
+                                "<a href='<%=issues_or_forum[:get_link].call(unread_notification_status.parent)%>'>" + 
+                                    "(<%=unread_notification_status.count%>) #<%=unread_notification_status.parent.id%> <%=unread_notification_status.parent.subject%>" +
+                                "</a>" + 
+                            "</li>"
+                <%end%>
+                html += "</ul>";
+                $("#notifications-placeholder").append(html);
+            <%end%>
+        <%end%>
+    <%end%>
+<%end%>
+
+<%# ============= Forum messages read/unread indicators on a topic page ============= %>
+<%
+    current_controller = controller.controller_name
+    current_action = controller.action_name
+%>
+<% if User.current.logged? && current_controller == 'messages' && current_action == 'show' %>
+    <%= javascript_tag do%>
+        var messageDOMNode = $("#content").children(".contextual, h2, .message").wrapAll("<div id='topic-wrapper' class='forum-topic'></div>").parent();
+        renderPluginForMessage("<%= @topic.id %>", <%= @topic.read_by_users.html_safe %>,"<%= escape_javascript(@topic.content.to_s[0, 130]) %>", 
+                                    '<%=l(:label_time_ago, :time => time_tag(@topic.created_on).html_safe).html_safe%>', messageDOMNode, "h2", ".message", "update_message_read_by_users",
+                                    ...<%= raw(getUsersReadStatus(@topic.read_by_users).to_json)%>);
+    <% end %>
+    <% if @replies != nil %>
+        <% @replies.each do |message| %>
+            <%= javascript_tag do%>
+                var messageDOMNode = $("#message-<%=message.id%>")
+                renderPluginForMessage("<%= message.id %>", <%= message.read_by_users.html_safe %>,"<%= escape_javascript(message.content.to_s[0, 130]) %>", 
+                                            '<%=l(:label_time_ago, :time => time_tag(message.created_on).html_safe).html_safe%>', messageDOMNode, ".reply-header", ".wiki", "update_message_read_by_users", 
+                                            ...<%= raw(getUsersReadStatus(message.read_by_users).to_json)%>);
```

Comment:
From what I understand here you cycle on each message and compute the users read.

`getUsersReadStatus` calls the db for each unknown user, but its internal cache is reset at the next call.

Maybe keep the users already read in a cache somewhere around here, and send it to `getUsersReadStatus` as well?

Meaning it will reach the db only once for a user no matter how many messages needs to be computed.

Reply:
I created a cache for the forum messages
But I didn't found a solution for the issues notes. Will search for one in another task:
https://redmine.xops-online.com/issues/33177